test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -1142,58 +1138,103 @@ arbitraryActionWithVars _ ctx _st = QC.frequency $ concat [
-
-        fromQueryResults ::
-             Var h (Either Model.Err (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob))))
-          -> Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))
-        fromQueryResults = mapGVar (\op -> OpRight `OpComp` OpQueryResults `OpComp` OpFromRight `OpComp` op)
-
-    withoutVars :: [(Int, Gen (Any (LockstepAction (ModelState h))))]
-    withoutVars = [
-          (5, Some . New (PrettyProxy :: PrettyProxy (k, v, blob)) <$> QC.arbitrary)
-        , (3, fmap Some $ Open @k @v @blob <$> genSnapshotName)
-        , (1, fmap Some $ DeleteSnapshot <$> genSnapshotName)
-        , (1, pure $ Some ListSnapshots)
+        lookupResultVars :: [Var h (V.Vector (R.LookupResult  v (WrapBlobRef h IO blob)))]
+        queryResultVars  :: [Var h (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob)))]
+
+        lookupResultVars = fromRight <$> findVars ctx Proxy
+        queryResultVars  = fromRight <$> findVars ctx Proxy
+
+    genUsedSnapshotName   = QC.elements usedSnapshotNames
+    genUnusedSnapshotName = QC.elements unusedSnapshotNames
+
+    usedSnapshotNames, unusedSnapshotNames :: [R.SnapshotName]
+    (usedSnapshotNames, unusedSnapshotNames) =
+      partitionEithers
+        [ if Map.member snapshotname (Model.snapshots st)
+            then Left  snapshotname -- used
+            else Right snapshotname -- unused
+        | name <- ["snap1", "snap2", "snap3" ]
+        , let snapshotname = fromJust (R.mkSnapshotName name)
         ]
 
-    withVars ::
-         Gen (Var h (Either Model.Err (WrapTableHandle h IO k v blob)))
-      -> [(Int, Gen (Any (LockstepAction (ModelState h))))]
-    withVars genVar = [
-          (2, fmap Some $ Close <$> (fromRight <$> genVar))
-        , (10, fmap Some $ Lookups <$> genLookupKeys <*> (fromRight <$> genVar))
-        , (5, fmap Some $ RangeLookup <$> genRange <*> (fromRight <$> genVar))
-        , (5, fmap Some $ NewCursor <$> QC.arbitrary <*> (fromRight <$> genVar))
-        , (10, fmap Some $ Updates <$> genUpdates <*> (fromRight <$> genVar))
-        , (10, fmap Some $ Inserts <$> genInserts <*> (fromRight <$> genVar))
-        , (10, fmap Some $ Deletes <$> genDeletes <*> (fromRight <$> genVar))
-        , (3, fmap Some $ Snapshot <$> genSnapshotName <*> (fromRight <$> genVar))
-        , (3, fmap Some $ Duplicate <$> (fromRight <$> genVar))
+    genActionsSession :: [(Int, Gen (Any (LockstepAction (ModelState h))))]
+    genActionsSession =
+        [ (5, fmap Some $ New  @k @v @blob PrettyProxy <$> QC.arbitrary)
+        | length tableVars <= 5 ] -- no more than 5 tables at once
+
+     ++ [ (3, fmap Some $ Open @k @v @blob <$> genUsedSnapshotName)
+        | not (null usedSnapshotNames) ]
+
+     ++ [ (1, fmap Some $ DeleteSnapshot <$> genUsedSnapshotName)
+        | not (null usedSnapshotNames) ]
+
+     ++ [ (1, fmap Some $ pure ListSnapshots)
+        | not (null tableVars) ] -- otherwise boring!
```

Comment:
Isn't `not (null usedSnapshotNames)` more interesting?

Reply:
I didn't want to completely exclude the case of there being no snapshots. I just want to avoid doing a boring list snapshots operations as the first action. Otherwise, at the start of the action sequence, when there are no tables, the only available actions are to make a new table or to list snapshots (with 1:1 chance). So by making it only available once there are tables, then the chance changes from 1:1 to more like 1:75 which is enough to not pick it all the time.